File: termux-shared/src/main/java/com/termux/shared/shell/ArgumentTokenizer.kt
package com.termux.shared.shell

/**
 * Tokenizes a command string into arguments, handling quotes and escapes.
 * Ported from DrJava (Rice University) — see original license in git history.
 * Pure Kotlin port matching the Rust implementation in terminal-emulator/src/main/rust/src/argument_tokenizer.rs.
 */
object ArgumentTokenizer {

    private const val NO_TOKEN = 0
    private const val NORMAL = 1
    private const val SINGLE_QUOTE = 2
    private const val DOUBLE_QUOTE = 3

    @JvmStatic
    fun tokenize(arguments: String): List<String> = tokenize(arguments, stringify = false)

    @JvmStatic
    fun tokenize(arguments: String, stringify: Boolean): List<String> {
        val args = mutableListOf<String>()
        val buf = StringBuilder()
        var escaped = false
        var state = NO_TOKEN
        val chars = arguments.toCharArray()
        var i = 0

        while (i < chars.size) {
            val c = chars[i]
            if (escaped) {
                escaped = false
                buf.append(c)
                i++
                continue
            }
            when (state) {
                SINGLE_QUOTE -> {
                    if (c == '\'') state = NORMAL else buf.append(c)
                }
                DOUBLE_QUOTE -> {
                    when {
                        c == '"' -> state = NORMAL
                        c == '\\' -> {
                            i++
                            if (i < chars.size) {
                                val next = chars[i]
                                if (next == '"' || next == '\\') buf.append(next)
                                else { buf.append(c); buf.append(next) }
                            } else buf.append(c)
                        }
                        else -> buf.append(c)
                    }
                }
                NO_TOKEN, NORMAL -> {
                    when (c) {
                        '\\' -> { escaped = true; state = NORMAL }
                        '\'' -> state = SINGLE_QUOTE
                        '"' -> state = DOUBLE_QUOTE
                        else -> {
                            if (!c.isWhitespace()) {
                                buf.append(c); state = NORMAL
                            } else if (state == NORMAL) {
                                args.add(buf.toString()); buf.clear(); state = NO_TOKEN
                            }
                        }
                    }
                }
            }
            i++
        }

        when {
            escaped -> { buf.append('\\'); args.add(buf.toString()) }
            state != NO_TOKEN -> args.add(buf.toString())
        }

        return if (stringify) {
            args.map { "\"${escapeQuotesAndBackslashes(it)}\"" }
        } else {
            args
        }
    }

    private fun escapeQuotesAndBackslashes(s: String): String = buildString(s.length + 4) {
        for (c in s) {
            when (c) {
                '\\', '"' -> { append('\\'); append(c) }
                '\n' -> append("\\n")
                '\t' -> append("\\t")
                '\r' -> append("\\r")
                '\u0008' -> append("\\b")
                '\u000C' -> append("\\f")
                else -> append(c)
            }
        }
    }
}

File: terminal-emulator/src/main/rust/src/argument_tokenizer.rs
/// Shell argument tokenizer — state machine that splits a command string
/// into arguments, handling single quotes, double quotes, and backslash escapes.
/// Ported from ArgumentTokenizer.java (DrJava, Rice University).

#[derive(Clone, Copy, PartialEq)]
enum State {
    NoToken,
    Normal,
    SingleQuote,
    DoubleQuote,
}

/// Tokenize a shell command string into arguments.
/// Returns a Vec of Strings, one per token.
pub fn tokenize(input: &str) -> Vec<String> {
    tokenize_inner(input, false)
}

/// Tokenize with optional stringification (wraps each arg in double quotes
/// and escapes internal quotes/backslashes).
pub fn tokenize_stringify(input: &str) -> Vec<String> {
    tokenize_inner(input, true)
}

fn tokenize_inner(input: &str, stringify: bool) -> Vec<String> {
    let mut args: Vec<String> = Vec::new();
    let mut current = String::new();
    let mut escaped = false;
    let mut state = State::NoToken;

    let chars: Vec<char> = input.chars().collect();
    let len = chars.len();
    let mut i = 0;

    while i < len {
        let c = chars[i];
        if escaped {
            escaped = false;
            current.push(c);
            i += 1;
            continue;
        }

        match state {
            State::SingleQuote => {
                if c == '\'' {
                    state = State::Normal;
                } else {
                    current.push(c);
                }
            }
            State::DoubleQuote => {
                if c == '"' {
                    state = State::Normal;
                } else if c == '\\' {
                    // Look ahead: only escape " or \ inside double quotes
                    i += 1;
                    if i < len {
                        let next = chars[i];
                        if next == '"' || next == '\\' {
                            current.push(next);
                        } else {
                            current.push(c);
                            current.push(next);
                        }
                    } else {
                        current.push(c);
                    }
                } else {
                    current.push(c);
                }
            }
            State::NoToken | State::Normal => {
                match c {
                    '\\' => {
                        escaped = true;
                        state = State::Normal;
                    }
                    '\'' => {
                        state = State::SingleQuote;
                    }
                    '"' => {
                        state = State::DoubleQuote;
                    }
                    _ if c.is_whitespace() => {
                        if state == State::Normal {
                            args.push(current.clone());
                            current.clear();
                            state = State::NoToken;
                        }
                        // NoToken + whitespace = skip
                    }
                    _ => {
                        current.push(c);
                        state = State::Normal;
                    }
                }
            }
        }
        i += 1;
    }

    // Handle trailing state
    if escaped {
        current.push('\\');
        args.push(current);
    } else if state != State::NoToken {
        args.push(current);
    }

    if stringify {
        for arg in &mut args {
            let escaped_arg = escape_quotes_and_backslashes(arg);
            *arg = format!("\"{}\"", escaped_arg);
        }
    }

    args
}

fn escape_quotes_and_backslashes(s: &str) -> String {
    let mut result = String::with_capacity(s.len() + 4);
    for c in s.chars() {
        match c {
            '\\' | '"' => {
                result.push('\\');
                result.push(c);
            }
            '\n' => result.push_str("\\n"),
            '\t' => result.push_str("\\t"),
            '\r' => result.push_str("\\r"),
            '\x08' => result.push_str("\\b"),
            '\x0C' => result.push_str("\\f"),
            _ => result.push(c),
        }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_args() {
        assert_eq!(tokenize("a b c"), vec!["a", "b", "c"]);
    }

    #[test]
    fn single_quotes() {
        assert_eq!(tokenize("a 'b c' d"), vec!["a", "b c", "d"]);
    }

    #[test]
    fn double_quotes() {
        assert_eq!(tokenize(r#"a "b c" d"#), vec!["a", "b c", "d"]);
    }

    #[test]
    fn backslash_escape() {
        assert_eq!(tokenize(r"a\ b c"), vec!["a b", "c"]);
    }

    #[test]
    fn empty_input() {
        assert_eq!(tokenize(""), Vec::<String>::new());
        assert_eq!(tokenize("   "), Vec::<String>::new());
    }

    #[test]
    fn stringified() {
        let result = tokenize_stringify("a 'b c'");
        assert_eq!(result, vec!["\"a\"", "\"b c\""]);
    }
}
